src/Util.h

```diff
@@ -321,6 +321,19 @@ using make_index_sequence = make_integer_sequence<size_t, N>;
 
 #endif
 
+// Helpers for timing blocks of code. Put 'TICK;' at the start and
+// 'TOCK;' at the end. Timing is reported at the tock via
+// debug(0). The calls can be nested and will pretty-print
+// appropriately. Took this idea from matlab via Jon Barron.
+//
+// Note that this uses global state internally, and is not thread-safe
+// at all. Only use it for single-threaded debugging sessions.
+
+void tick_impl(const char *file, int line);
```

Comment:
The idea is fine, and seems fairly standard in timing abstractions, but the naming is more cute than productive. First off, we just cannot declare macros that don't have a "HALIDE_" prefix. Just not a good idea. (Specifically this ends up in Halide.h right? They get undefined at the end by Halidefooter.h, but there isn't anything dealing with incoming definitions.)

Second, if I read the documentation correctly, TICK is "start_timing" and TOCK is "end_timing_and_report" which is not what I would expect TICK and TOCK to mean. I would expect TICK and TOCK to be sequencers that bounce between two mostly equivalent things, such as timing execution and then delay until the next execution.

How about HALIDE_TIMER_START and HALIDE_TIMER_END?

Also don't you want to use fully qualified names in the macro expansion?

Reply:
The names are cute, but semi-standardized - I'm just using what matlab/octave use with exactly the same meaning. Not sure how many other systems do the same. 

The idea of these is to make it really easy to drop in temporarily during a performance debugging session, so short names are essential to this being productive.

Other macros we define include user_error, user_warning, internal_error, etc. There doesn't seem to be any protection against these already being defined before Halide.h is included. Perhaps there should be.

Reply:
Actually looks like matlab is TIC/TOC, not TICK/TOCK.

HALIDE_TIC/HALIDE_TOC?

Reply:
Also note that these should never appear in checked-in code, so unlike most code, we should be optimizing for writing, not reading.

Reply:
And we have had people complain about user_error etc. We should change those but it is a lot of work. The names being standardized makes things worse because it means some other package will use the same names but with different definitions.

Reply:
I'll also argue that with identifier completion, the productivity difference between TIC/TOC and HALIDE_TIC/HALIDE_TOC is minimal. Even if you have to type them, the win is in not having to make any decisions about what arguments to pass, etc. not the really short number of characters.

We really need an iron clad rule that Halide.h does not define any macros that are not prefixed with halide_ or HALIDE_.

Reply:
I think we need a different solution, because halide_internal_assert is going to slightly discourage writing assertions and make it slightly more annoying to read them. Surely there's a mechanism for this. Copy-pasting from gitter, does this work?

// top of Halide.h
ifdef user_assert
define HALIDE_SAVE_user_assert user_assert
endif

...

define user_assert ...

...

// bottom of Halide.h
undef user_assert
ifdef HALIDE_SAVE_user_assert
define user_assert HALIDE_SAVE_user_assert
undef HALIDE_SAVE_user_assert

We have control over what goes into Halide.h vs what stays in internal headers. Another possible solution is to just not let these macros into Halide.h

Reply:
Yes, one can save the macros, at least for ones that don't take arguments. But it is a hack and as one might expect, there are a lot of ways to lose. I'd rather not create more issues that have to be hacked around later.

For e.g. user_assert, I expect we want to make Halide.h have "halide_user_assert" and then a way that inside the Halide build itself to get:

    #define user_assert halide_user_assert

in scope. (Easiest was to do that is via an #ifdef on another macro that the build system defines.)

Reply:
That sounds like an ideal solution. I don't want to do it in this PR (which is really about skip_stages), so I'll just add halide prefixes to the new macros here.

Reply:
Yeah, just prefixing these macros is all I'm asking for. (The both halide_user_assert and maybe user_assert solution might require providing the macro bodies twice to get the __FILE__ and __LINE__ info right. There's supposedly a way to pass caller location more cleanly being added for e.g. C++20.)

Reply:
Also, that means the tests can use these macros, just with the halide_ prefix.